Added BranchesManagement beginning

piotrkow95 opened the pull request:
-Added branches list with employees and currently available cars in each branch with html view
-Notification about 1 or 2 (less than 3) cars available
-Failed attempt to check manager in each branch (commented)

piotrkow95 committed:
-Added branches list with employees and currently available cars in each branch
-Notification about 1 or 2 (less than 3) cars available
-Failed attempt to check manager in each branch (commented)

Mbojanow commented on src/main/java/pl/sdacademy/carrental/controllers/BranchController.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+
+@Controller
+@RequestMapping("/list-branches")
```
path jest specyficzny dla akcji, czyli powienien znaleźć się przy endpoincie.

MNowicki87 replied:
Jeśli dobrze rozumiem, to ten Controller będzie wykorzystywany do zarządzania wypożyczalnią/oddziałem - jeśli tak, to powinniśmy już na tym etapie wsadzać to w wydzielone miejsce w urlu - np. `"/admin/**"`

Mbojanow commented on src/main/java/pl/sdacademy/carrental/controllers/BranchController.java:
```diff
@@ -0,0 +1,40 @@
+    @GetMapping
+    public String listAllBranches(final Model model) {
+        final List<Branch> allBranches = branchService.getAll();
+        final boolean notification = carService.availableOnlyOneOrTwoCars();
```
brak czasownika w nazwie metody. Może zmienić na 'isLessThanThreeCarsAvailable'? Poza tym nagle informacja o dostępności zmienia się na informacje o notyfikacji, tzn lepiej byłoby stworzyć metodę coś jak np. 'shouldNotifyAboutLowCarAvailability' i pod spodem wykonać logikę z countem < 3.

piotrkow95 replied:
Zmienię nazwę metody za chwilę na odpowiedniejszą :D Jeszcze uczę się obsługiwać pull requesty. Będę musiał wykonać całkowicie nowego pull requesta z tą zmienioną nazwą metody?

MNowicki87 replied:
wystarczy, że zacommitujesz zmiany i wypchniesz na brancha, który ma być zmergowany - commit zostanie uwzględniony w PR

Mbojanow commented on src/main/java/pl/sdacademy/carrental/controllers/BranchController.java:
```diff
@@ -0,0 +1,40 @@
+//        final List<Employee> managerList = employeeService.getManager();
+        model.addAttribute("branchList", allBranches);
+//        model.addAttribute("managerList",managerList);
+        model.addAttribute("oneOrTwoCarsNotification",notification);
```
formatowanie

Mbojanow commented on src/main/java/pl/sdacademy/carrental/services/CarService.java:
```diff
@@ -26,4 +27,12 @@ public CarService(final CarRepository carRepo) {
+   }
+
+   public boolean availableOnlyOneOrTwoCars(){
+      return countAvailableCars() == 1 || countAvailableCars() == 2;
```
ratuje Cię 1st level cache przez wykonywaniem zapytania wielokrotnie :).
Można zmienić na countAvailableCars() < 3 i ta 3 powinna mieć nazwę (magic number in da code)

piotrkow95 replied:
Co do tego to miałem wątpliwości, czy mogę tak zrobić, ponieważ informacja chyba miała zawiadamiać, że samochody jeszcze są ale tylko 1 lub 2. A " < 3 " zawierałoby też przypadek zerowy

Mbojanow replied:
ok, to wtedy count między 0 a 3. Tak czy inaczej nie wywołuj metody `countAvailableCars` 2 razy (co jeśli trwa ona długo i nie jest cachowana? Akurat w tym przypadku hibernate Cię ratuje)

MNowicki87 replied:
> Można zmienić na countAvailableCars() < 3 i ta **3 powinna mieć nazwę** (magic number in da code)

Proponuję stałą nazwać raczej ogólnie - np. MINIMUM_CAR_COUNT. Metodę sprawdzającą również - bez liczb, bo te w przyszłości mogą być zmieniane (wraz z rosnącym ruchem może się okazać, że trzeba podnieść próg).

Można również przekazać liczbę metodą count.. (domyślam się, że jest publiczna, bo będzie wykorzystywana wyżej) i tam złapać ją do zmiennej (i pewnie dodać do modelu), i wtedy rzucić jakimś powiadomieniem, że jest za mało.

Mbojanow commented on src/main/resources/templates/branches-list.html:
```diff
@@ -0,0 +1,29 @@
+            <p></p>
+            <p th:text="${branch.getCarsOnHand()}"></p>
+        </li>
+        <li th:each="branch : ${oneOrTwoCarsNotification}">
```
ta zmienna w modelMapie jest boolean -> nie ma sensu się przez nią iterować

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/services/CarService.java:
```diff
@@ -26,4 +27,12 @@ public CarService(final CarRepository carRepo) {
    }
+
+   public long countAvailableCars(){
+      return getAvailable().size();
```
Metoda `getAvailable()` wyciąga wszystkie samochody, które są aktualnie dostępne globalnie (w całej firmie, niezależnie od oddziału).
Po ilość chyba lepiej strzelać do bazy count'em - tych samochodów może trochę być i nie ma potrzeby pobierać z bazy wszystkich danych tylko po to, żeby je policzyć w aplikacji (tym bardziej, że poza liczeniem nie jest wykonywana żadna logika).

piotrkow95 replied:
Na swoją obronę w tym przypadku mogę powiedzieć, że inteliJ przy refaktoryzacji po próbie commitu proponował zamienić na size() i w sumie nie spędziłem chwili na zastanowieniu się czy to na pewno lepsze rozwiązanie. Wrócę do counta

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/repositories/EmployeeRepository.java:
```diff
@@ -0,0 +1,13 @@
+public interface EmployeeRepository extends JpaRepository<Employee,Long> {
+
+
+//    List<Employee> findEmployeesByRole(EmployeeRole role);
```
nie komentuj metod :-) jeśli będziesz z niej korzystać, to zostaw _uncommented_, a jeśli jeszcze nie wiesz, czy się przyda, to po prostu usuń - w razie potrzeby 1 linię dopiszesz w kilka sekund ;-)

piotrkow95 replied:
Zły nawyk z niechlujnego kodzenia we własnym podwórku, lubię komentować stare rozwiązania czy pomysły na przyszłość. Takie komentarze już się nie pojawią ;p

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/services/EmployeeService.java:
```diff
@@ -0,0 +1,26 @@
+
+//    public List<Employee> getManager(){
+//        return employeeRepo.findEmployeesByRole(EmployeeRole.MANAGER);
+//    }
```
-> to, co w EmployeeRepo

MNowicki87 reviewed (changes requested):
Komentarze przy kodzie

piotrkow95 committed:
-Added branches list with employees and currently available cars in each branch
-Notification about 1 or 2 (less than 3) cars available
-Failed attempt to check manager in each branch (commented)

Fixed according to reviews, comments deleted

MNowicki87 committed:
set car.currentStatus as EnumType.STRING

MNowicki87 committed:
mapped BranchController to /admin; set endpoint-specific url

oskprzym commented:
Właściwie wszystkie wątpliwości, jakie mogłem mieć, zostały tu już opisane/rozwiązane, zamykam PR.

oskprzym commented:
Coś się pomyliłem w obsłudze gita, otwieram i zostawiam póki co :)

MNowicki87 committed:
- BranchController now gets a Map<Branch, Integer> where Integer is a number of currently available cars in branch
- moved MINIMUM_CAR_COUNT from CarService to BranchService the check is done in the front
- added one more car to OnAppStartup to check if notification works properly

MNowicki87 committed:
too fast - forgot to remove CarService from BranchController constructor :-)

piotrkow95 committed:
Merge pull request #8 from SdaJavaGda32-teamCarRental/developPK_MN

Develop pk mn
